tests/DCPS/Presentation/main.cpp

```diff
@@ -8,350 +10,584 @@
-
-    FooDataWriter_var writer_i = FooDataWriter::_narrow(writer);
-    if (CORBA::is_nil(writer_i))
-    {
-      ACE_ERROR_RETURN((LM_ERROR,
-                        ACE_TEXT("%N:%l main()")
-                        ACE_TEXT(" ERROR: _narrow failed!\n")), -1);
+    if (CORBA::is_nil(writer_.in())) {
+      ACE_ERROR_RETURN((LM_ERROR, ACE_TEXT("%N:%l setup_test() ERROR: ")
+                        ACE_TEXT("create_datawriter failed!\n")), true);
     }
 
     // Block until Subscriber is associated
-    DDS::StatusCondition_var cond = writer->get_statuscondition();
-    cond->set_enabled_statuses(DDS::PUBLICATION_MATCHED_STATUS);
+    {
+      DDS::StatusCondition_var cond = writer_->get_statuscondition();
+      cond->set_enabled_statuses(DDS::PUBLICATION_MATCHED_STATUS);
+      DDS::WaitSet_var ws = new DDS::WaitSet;
+      ws->attach_condition(cond);
+      DDS::ConditionSeq conditions;
+      DDS::PublicationMatchedStatus matches = { 0, 0, 0, 0, 0 };
+      do {
+        rc = ws->wait(conditions, max_wait_time);
+        if (rc != DDS::RETCODE_OK) {
+          ACE_ERROR_RETURN((LM_ERROR, ACE_TEXT("%N:%l setup_test() ERROR: ")
+                            ACE_TEXT("wait failed: %C\n"),
+                            retcode_to_dds_string(rc).c_str()), true);
+        }
 
-    DDS::WaitSet_var ws = new DDS::WaitSet;
-    ws->attach_condition(cond);
+        rc = writer_->get_publication_matched_status(matches);
+        if (rc != ::DDS::RETCODE_OK) {
+          ACE_ERROR_RETURN((LM_ERROR, ACE_TEXT("%N:%l setup_test() ERROR: ")
+                            ACE_TEXT("Failed to get publication match status: %C\n"),
+                            retcode_to_dds_string(rc).c_str()), true);
+        }
+      } while (matches.total_count < (coherent ? 2 : 1));
+      ws->detach_condition(cond);
+    }
 
-    DDS::Duration_t timeout =
-      { DDS::DURATION_INFINITE_SEC, DDS::DURATION_INFINITE_NSEC };
+    return false;
+  }
 
-    DDS::ConditionSeq conditions;
-    DDS::PublicationMatchedStatus matches = { 0, 0, 0, 0, 0 };
-    do
-    {
-      if (ws->wait(conditions, timeout) != DDS::RETCODE_OK)
-      {
-        ACE_ERROR_RETURN((LM_ERROR,
-                          ACE_TEXT("%N:%l main()")
-                          ACE_TEXT(" ERROR: wait failed!\n")), -1);
+  ~Test()
+  {
+    DDS::ReturnCode_t rc;
+
+    if (!CORBA::is_nil(reader1_)) {
+      rc = subscriber_->delete_datareader(reader1_);
+      if (rc != DDS::RETCODE_OK) {
+        ACE_ERROR((LM_ERROR, ACE_TEXT("%N:%l ~Test() ERROR: ")
+          ACE_TEXT("delete reader1 failed: %C\n"),
+          retcode_to_dds_string(rc).c_str()));
       }
+    }
 
-      if (writer->get_publication_matched_status(matches) != ::DDS::RETCODE_OK)
-      {
-        ACE_ERROR_RETURN((LM_ERROR,
-                          ACE_TEXT("%N:%l main()")
-                          ACE_TEXT(" ERROR: Failed to get publication match status!\n")), -1);
+    if (!CORBA::is_nil(reader2_)) {
+      reader2_->set_listener(0, 0);
```

Comment:
Does it need to remove a listener if the object is being deleted on the next line?

Reply:
I suppose not